### DISCUSSION TITLE
up-follow doesn't fire in a "Bootstrap Table" cell

### DISCUSSION THREAD

**Discussion by patosan** [unpoly/unpoly#455] [Q&A] [answered]
Hi all, I'm brand new to Unpoly and Javascriptland. I'm using Unpoly with [Bootstrap Table](https://bootstrap-table.com/) (a jquery plugin). For some reason, `up-follow` and `up-preload` don't fire when a link is inside a Bootstrap Table's cell. These HTML attributes fire just fine outside the table.

I've created a [Glitch Project](https://glitch.com/edit/#!/tundra-morning-wanderer) to demonstrate the problem. Below are my site wide settings (scripts.js), which mandates up-follow and up-preload on all links.

```
up.link.config.followSelectors.push("a[href]");
up.link.config.preloadSelectors.push("a[href]");
up.link.config.preloadDelay = 10;
up.form.config.submitSelectors.push(["form"]);
```

These settings work great for the Home and About links in the project. Those two links behave exactly as they should. However, links inside the table's cells will not fire `up-follow` or `up-preload`. On the first row of the table, I even tried manually setting  `up-follow`, `up-preload`, and `up-transition="move-left"` attributes on the first row's `<a>` tag. Of those three HTML attributes, only up-transition fires successfully.

I'm not certain if I'm doing something wrong or missing something that would be obvious to someone with more JavaScript experience. I've spent a lot of time trying to figure this out on my own, but I don't know enough about JavaScript to dig deeper into the problem. Any help would be really appreciated!

**Comment by adam12** [comment, marked as answer]
I suspect that Bootstrap Table is manipulating the DOM inside the Unpoly compiler, and Unpoly doesn't see the changes. Normally you can call `up.hello(element)` from an event, which will have Unpoly re-recognize those elements again.

I tried to come up with a fix for you, but I am not familiar with Bootstrap Table and have only limited time today. This is where I was going:

```js
 $("#html-data").bootstrapTable("destroy").bootstrapTable({
    onPostBody: () => up.hello(element.querySelector('tbody'))
 });
```

If you `up.hello` on the `#html-data` element, you'll get an infinite loop, so you would need to trigger it on the element that doesn't trigger the compiler (removing the `id` on first compiler pass might work). I tried selecting the `tbody` of the table, but table child elements are special cases in browsers, and may not parse correctly without a parent `<table>`.

Perhaps you could look at either:

a) getting rid of the infinite loop by deleting the `#html-data` ID when the compiler first runs. Then `up.hello(element)` in the hook as I show.

b) move the `#html-data` to a wrapper of the `<table>`, and then `up.hello(element.querySelector('table'))`, which would allow Unpoly to look at the entire table without risk of an infinite loop.

**Reply by adam12** [reply]
Another thought might be to set an attribute on the element at the start compiler if it's not set. If it is set already, skip the setup.

**Comment by patosan** [comment]
Adam12,

Thank you for spending precious weekend time helping me with this problem! I was starting to despair that I wouldn't be able to fix it. I appreciate it more than you can ever know!

The code you provided gets up-follow and up-preload firing again! One small issue I can see is if the table's header needs to be reloaded by Unpoly, which is addressed by adding an up-hello call for "thead":
```
up.compiler("#html-data", function (element) {
  $("#html-data").bootstrapTable("destroy").bootstrapTable({
    onPostBody: () => {
      up.hello(element.querySelector("thead"));
      up.hello(element.querySelector("tbody"));
    }
  });
});
```
I'm uncertain why your  code with my small addition above isn't more or less sufficient? The only hole I can see in the above code is if Unpoly needed to reload the `<table>` element for some reason. I can't see this ever occurring in my project (famous last words 😅️). That said, I'm 100% new to JavaScript, so it's _very_ possible I'm missing something.

**Reply by adam12** [reply]
> One small issue I can see is if the table's header needs to be reloaded by Unpoly,

There are other events that Bootstrap Table emits. You might need to add another one to capture changes there too. I think I saw one for the header.

> I'm uncertain why your code with my small addition above isn't more or less sufficient?

If this is working for you, I'd say ship it :shipit:

**Comment by patosan** [comment]
Awesome, thanks again for your help! I'm loving all the cool toys in Unpoly. It's super powerful and very well thought out!
